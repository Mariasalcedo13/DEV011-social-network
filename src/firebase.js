import {
  deleteDoc,
  query,
  where,
  getFirestore,
  collection,
  getDocs,
  addDoc,
  doc,
  getDoc,
  updateDoc,
  onSnapshot,
} from 'firebase/firestore';
import {
  getAuth,
  GoogleAuthProvider,
  signInWithPopup,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
} from 'firebase/auth';
import {
  ref,
  uploadBytes,
  getDownloadURL,
  getStorage,
} from 'firebase/storage';
import { initializeApp } from 'firebase/app';

// import renderCreateAccount from './register.js';
// import { posts } from './post.js';
// import navigateTo from './main';

const firebaseConfig = {
  apiKey: 'AIzaSyA_nNPVRwXqmgLlxdYL4NmJwiItX9t2D5E',
  authDomain: 'social-network-c61c9.firebaseapp.com',
  projectId: 'social-network-c61c9',
  storageBucket: 'social-network-c61c9.appspot.com',
  messagingSenderId: '496904934051',
  appId: '1:496904934051:web:349b4f181faf09491c2516',
  measurementId: 'G-9MSK8FV9VP',
};
// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const firestore = getFirestore(app);
const googleProvider = new GoogleAuthProvider();
googleProvider.setCustomParameters({
  prompt: 'select_account',
});
const storage = getStorage(app);
export {
  app, firestore, googleProvider, auth,
};

// Para crear o registrar usuarios
export function createUser(email, password) {
  return new Promise((resolve, reject) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        console.log('Registro exitoso', userCredential);
        const user = userCredential.user;
        resolve({ message: 'success', email: user.email });
      })
      .catch((error) => {
        console.error(
          'Error al registrarse:',
          error.code,
          error.message,
          error.serverResponse,
        );
        error.email = email;
        reject(error);
      });
  });
}

// Para iniciar sesion o ingresar
export function login(email, password) {
  return new Promise((resolve, reject) => {
    signInWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        console.log('Inicio sesion exitoso', userCredential);
        const user = userCredential.user;
        resolve({ message: 'success', email: user.email });
      })
      .catch((error) => {
        console.error(
          'Error al iniciar sesion:',
          error.code,
          error.message,
          error.serverResponse,
        );
        error.email = email;
        reject(error);
      });
  });
}

// function save
// export function saveTask(title, description, imageInput) {

//   const imageName = `${Date.now()}_${imageInput.name}`;
//   const storageRef = ref(storage, `/img/${imageName}`);
//   uploadBytes(storageRef, imageInput).then((snapshot) => {
//     console.log('Uploaded a blob or file!');
//   });
//   console.log(storageRef);
//   let image = {};
//   const postCollection = collection(firestore, 'post');
//   addDoc(postCollection, {
//     title,
//     description,
//     image,
//     likes: 0,
//   })
//     .then((docRef) => {
//       console.log('Documento guardado con ID:', docRef.id);
//     })
//     .catch((error) => {
//       console.error('Error al guardar el documento:', error);
//     });
// }
// No necesitas el async/await aquí
export async function saveTask(title, description, imageFile) {
  console.log(imageFile);
  if (!imageFile) {
    const postCollection = collection(firestore, 'post');
    // Guarda la información del post en la base de datos
    try {
      await addDoc(postCollection, {
        title,
        description,
        likes: 0,
      });
    } catch (e) {
      console.error(e);
    }
  } else {
    try {
      const imageName = `${Date.now()}_${imageFile.name}`;
      const storageRef = ref(storage, `/img/${imageName}`);
      // Sube la imagen al almacenamiento y obtiene la URL de descarga
      const imageUrl = await uploadBytes(storageRef, imageFile)
        .then(() => getDownloadURL(storageRef));
      console.log('URL de la imagen:', imageUrl);
      const postCollection = collection(firestore, 'post');
      // Guarda la información del post en la base de datos
      const docRef = await addDoc(postCollection, {
        title,
        description,
        imageUrl,
        likes: 0,
      });
      console.log('Documento guardado con ID:', docRef.id);
    } catch (error) {
      console.error('Error al guardar el documento:', error);
      throw error; // Puedes manejar el error según tus necesidades
    }
  }
}

// funcion para registro con google
export function GoogleRegister(navigateTo) {
  signInWithPopup(auth, googleProvider)
    .then((result) => {
      const user = result.user;
      console.log('Usuario autenticado con Google:', user);
      navigateTo('/posts');
    })
    .catch((error) => {
      console.error('Error al autenticar con Google:', error.message);
    });
}

// funcion para likes
export function handleLike(postId, userId, callback) {
  const likesCollection = collection(firestore, 'likes');
  const likeButton = document.querySelector(
    `.likeButton[data-post-id="${postId}"]`,
  );
  const likeQuery = query(
    likesCollection,
    where('postId', '==', postId),
    where('userId', '==', userId),
  );

  // Ejecuta la consulta para verificar si el usuario ya ha dado like
  getDocs(likeQuery)
    .then((querySnapshot) => {
      // Verifica si no hay likes existentes del usuario para esta publicación
      if (querySnapshot.empty) {
        const postRef = doc(firestore, 'post', postId);

        // Obtiene la información actual de la publicación
        getDoc(postRef)
          .then((docSnapshot) => {
            if (docSnapshot.exists()) {
              const currentLikes = docSnapshot.data().likes;
              const updatedLikes = currentLikes + 1;
              likeButton.style.background = 'green';
              // Actualiza la cantidad de likes en la publicación
              updateDoc(postRef, { likes: updatedLikes })
                .then(() => {
                  console.log('Like registrado con éxito.');

                  // Agrega un nuevo documento de like
                  addDoc(likesCollection, { postId, userId })
                    .then(() => {
                      console.log('Nuevo like registrado con éxito.');
                      callback(); // Llama al callback después de manejar el like
                    })
                    .catch((error) => {
                      console.error('Error al agregar nuevo like:', error);
                    });
                })
                .catch((error) => {
                  console.error('Error al actualizar likes:', error);
                });
            }
          })
          .catch((error) => {
            console.error('Error al obtener documento de publicación:', error);
          });
      } else {
        // El usuario ya ha dado like antes, manejar según sea necesario
        const likeDoc = querySnapshot.docs[0]; // Obtén el documento de like existente
        const likeId = likeDoc.id;

        // Elimina el like existente
        deleteDoc(doc(likesCollection, likeId))
          .then(() => {
            console.log('Like eliminado con éxito.');

            // Obtiene la información actual de la publicación
            const postRef = doc(firestore, 'post', postId);
            getDoc(postRef)
              .then((docSnapshot) => {
                if (docSnapshot.exists()) {
                  const currentLikes = docSnapshot.data().likes;
                  const updatedLikes = currentLikes - 1;

                  // Actualiza la cantidad de likes en la publicación
                  updateDoc(postRef, { likes: updatedLikes })
                    .then(() => {
                      console.log('Like actualizado con éxito.');
                      callback(); // Llama al callback después de manejar el unlike
                    })
                    .catch((error) => {
                      console.error('Error al actualizar likes:', error);
                    });
                }
              })
              .catch((error) => {
                console.error(
                  'Error al obtener documento de publicación:',
                  error,
                );
              });
          })
          .catch((error) => {
            console.error('Error al eliminar like:', error);
          });
      }
    })
    .catch((error) => {
      console.error('Error al realizar la consulta de likes:', error);
    });
}

// para eliminar post
export function deletePost(postId) {
  const postCollection = collection(firestore, 'post');
  const postDocRef = doc(postCollection, postId);
  deleteDoc(postDocRef)
    .then(() => {
      console.log('Post eliminado con éxito.');
    })
    .catch((error) => {
      console.error('Error al borrar post:', error);
    });
}

// para editar los post
export function editPost(postId, updatedTitle, updatedDescription) {
  const postRef = doc(firestore, 'post', postId);

  return updateDoc(postRef, {
    title: updatedTitle,
    description: updatedDescription,
  });
}

// funcion cerrar sesion
export function logOut(navigateTo) {
  signOut(auth)
    .then(() => {
      // Sign-out successful.
      console.error('sesion cerrada');
      navigateTo('/login');
    })
    .catch((error) => {
      console.error('Error al cerrar sesion:', error.code, error.message);
      // An error happened.
    });
}

export function initializeAuth(setupPost, navigateTo) {
  onAuthStateChanged(auth, (user) => {
    if (user) {
      // Usuario autenticado, puedes acceder a la colección de 'post'
      console.log(
        'User authenticated:',
        auth.currentUser.uid,
        'email:',
        user.email,
      );
      const userPostsCollection = collection(firestore, 'post');
      onSnapshot(userPostsCollection, (snapshot) => {
        const postSnap = [];
        snapshot.forEach((docu) => {
          postSnap.push(docu);
        });
        setupPost(postSnap);
      });
    } else {
      console.log('Usuario no autenticado');
      // logOut();
      alert('Usuario no autenticado');
      navigateTo('/login');
      // navigateTo('/login');
      // backgroundLayer.style.opacity = 0.0;
      // Resto del código de la lógica que quieras tras un usuario no autenticado
    }
  });
}
