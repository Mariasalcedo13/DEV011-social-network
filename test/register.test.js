/**
 * @jest-environment jsdom
 */

import renderCreateAccount from '../src/register.js';
import * as firebase from '../src/firebase.js';

describe('renderCreateAccount', () => {
  it('Should be a function', () => {
    expect(typeof renderCreateAccount).toBe('function');
  });
  it('Have a button to return', () => {
    const DOM = document.createElement('div');
    DOM.append(renderCreateAccount());
    const haveAbutton = DOM.querySelector('#return');
    expect(haveAbutton).not.toBe(undefined);
  });
  it('Have a button to register with Google', () => {
    const DOM = document.createElement('div');
    DOM.append(renderCreateAccount());
    const haveAbuttonGoogle = DOM.querySelector('.buttonRegister');
    expect(haveAbuttonGoogle).not.toBe(undefined);
  });
  it('After click button return call function navigateTo with /', () => {
    const DOM = document.createElement('div');
    const navigateTo = jest.fn();
    DOM.append(renderCreateAccount(navigateTo));
    const haveAbuttonReturn = DOM.querySelector('#return');
    haveAbuttonReturn.click();
    expect(navigateTo).toHaveBeenLastCalledWith('/');
  });
  it('Snapshot of register', () => {
    const DOM = document.createElement('div');
    DOM.append(renderCreateAccount());
    expect(DOM).toMatchSnapshot();
  });
});

describe('Button register', () => {
  it('test of click button register', () => {
    const DOM = document.createElement('div');
    document.body.appendChild(DOM);
    DOM.append(renderCreateAccount());
    jest.spyOn(firebase, 'createUser').mockImplementation(() =>
    // eslint-disable-next-line
      Promise.resolve({
        message: 'success',
        email: 'giselle@lopez.com',
      }));

    // jest.spyOn(auth, 'login').mockImplementation(() => Promise.resolve({
    // message: 'success', email: 'giselle@holaLopez.com' }))
    const buttonRegister = DOM.querySelector('#continue');
    const email = DOM.querySelector('#emailRegister');
    email.value = 'giselle@holaLopez.com';
    const password = DOM.querySelector('#passwordRegister');
    password.value = 'hola123';
    // await Promise.resolve();
    buttonRegister.click();
    expect(firebase.createUser).toHaveBeenCalledTimes(1);
    // loginSpy.mockRestore();
  });
});
