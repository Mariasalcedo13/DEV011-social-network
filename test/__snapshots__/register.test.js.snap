// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderCreateAccount Snapshot of register 1`] = `
<div>
  <div
    class="homepage1"
  >
    <header
      class="headerRegister"
    >
      <h2>
        Para crear una nueva cuenta, ingresa tus datos.
      </h2>
      <img
        src="img/evolucion.png"
        style="width: 40%; height: auto;"
      />
    </header>
    <div
      class="Inputcontainer"
    >
      <h4
        class="h4Register"
      >
        Nombres completos:
      </h4>
      <input
        class="input"
        type="text"
      />
      <h4
        class="h4Register"
      >
        Correo Electrónico:
      </h4>
      <input
        class="input"
        id="emailRegister"
        type="email"
      />
      <h4
        class="h4Register"
      >
        Contraseña:
      </h4>
      <input
        class="input"
        id="passwordRegister"
        type="password"
      />
      <span
        id="answerPass"
      />
      <button
        class="buttonRegister"
        id="continue"
      >
        Continuar
      </button>
      <h4
        id="or"
      >
        o
      </h4>
      <button
        class="buttonRegister"
        id="google"
      >
        <img
          class="imgGoogle"
          src="img/googleongpng.webp"
        />
        <h2>
          Registrarse con Google
        </h2>
      </button>
      <button
        class="buttonRegister"
        id="return"
      >
        Volver
      </button>
    </div>
  </div>
</div>
`;
